### PR TITLE
fix: export link regex for use in edit message

### PR DESCRIPTION
**Pull request by s-alves10**
### Fixed Issues
$ https://github.com/Expensify/App/issues/18911

# Tests
1. What unit/integration tests cover your change? What autoQA tests cover your change?
ExpensiMark-HTML-test.js. Nothing added because there is no regex change.
2. What tests did you perform that validates your changed worked?
Make sure that all unit-tests are passed.

# QA
1. What does QA need to do to validate your changes?
This PR has no regex change. Can verify if all unit-tests are passed.
2. What areas to they need to test for regressions?
Nothing.

### CODE REVIEW

**Comment by s-alves10** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by s-alves10** [comment]
recheck

**Comment by s-alves10** [comment]
@MariaHCD 
Please review this PR asap. This PR needs to be merged in order to complete [this PR](https://github.com/Expensify/App/pull/19216) for [issue 18911](https://github.com/Expensify/App/issues/18911)

**Comment by s-alves10** [comment]
> @s-alves10 your current change itself looks good. But how will you solve this regression - [Expensify/App#18911 (comment)](https://github.com/Expensify/App/issues/18911#issuecomment-1552806599)?

We need to share the same regex in editing comment with `expensify-common`. When we send the comment, we escape the string first and pass the rules. I did the same. When checking if a link exists, escape the editing comment and run the regex. This works fine.

image

<details>
<summary>result</summary>

https://github.com/Expensify/expensify-common/assets/126839717/932f9552-6d88-4e6c-b975-02e2b7fd4551



</details>

File: lib/ExpensiMark.js [hunk 1]
```diff
@@ -3,6 +3,8 @@ import Str from './str';
 import {MARKDOWN_URL_REGEX, LOOSE_URL_REGEX, URL_REGEX} from './Url';
 import {CONST} from './CONST';
 
+const MARKDOWN_LINK_REGEX = `\\[([^\\][]*(?:\\[[^\\][]*][^\\][]*)*)]\\(${MARKDOWN_URL_REGEX}\\)(?![^<]*(<\\/pre>|<\\/code>))`;
```
[the remark this answers is not in the document]
Comment (s-alves10) [reply lib/ExpensiMark.js:6]: @aimane-chnaif 
I intentionally place this constant in ExpensiMark.js. This is not the URL regex, is the regex for parsing markdown URL.
Should I still move this regex to Url.js?

**Comment by s-alves10** [comment]
> can you explain why escape should work?

@aimane-chnaif 
When we escape the string, `<` and `>` are escaped to `&lt;` and `&gt;`. That's the reason.

File: lib/ExpensiMark.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (s-alves10) [reply lib/ExpensiMark.js]: @aimane-chnaif 
Moved

**Comment by s-alves10** [comment]
@neil-marcellini @MariaHCD 
This PR needs to be merged in order to complete [this PR](https://github.com/Expensify/App/pull/19216) for issue [18911](https://github.com/Expensify/App/issues/18911).

Will you review the PR?

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Sorry to keep changing my recommendations, I want to make sure this works as well as possible and stays consistent in the future.

How about we just move [these two functions from App](https://github.com/Expensify/App/blob/0ec61bf1ca783ee0e30fbbc3db7285a33e96744a/src/libs/actions/Report.js#L900-L924) to expensify-common, then we can simply use them in App?

File: lib/ExpensiMark.js [hunk 2]
```diff
@@ -73,10 +73,7 @@ export default class ExpensiMark {
             {
                 name: 'link',
                 process: (textToProcess, replacement) => {
-                    const regex = new RegExp(
-                        `\\[([^\\][]*(?:\\[[^\\][]*][^\\][]*)*)]\\(${MARKDOWN_URL_REGEX}\\)(?![^<]*(<\\/pre>|<\\/code>))`,
-                        'gi',
-                    );
+                    const regex = new RegExp(MARKDOWN_LINK_REGEX, 'gi');
```
[the remark this answers is not in the document]
Comment (s-alves10) [reply lib/ExpensiMark.js:76]: I think `gi` flag is intentional not to parse multiline link. In App, we'll be using `gi` flags as well.

File: lib/Url.js [hunk 3]
```diff
@@ -16,6 +16,7 @@ const LOOSE_URL_REGEX = `((${LOOSE_URL_WEBSITE_REGEX})${URL_PATH_REGEX}(?:${URL_
 
 
 const MARKDOWN_URL_REGEX = `(${LOOSE_URL_REGEX}|${URL_REGEX})`;
+const MARKDOWN_LINK_REGEX = `\\[([^\\][]*(?:\\[[^\\][]*][^\\][]*)*)]\\(${MARKDOWN_URL_REGEX}\\)(?![^<]*(<\\/pre>|<\\/code>))`;
```
[the remark this answers is not in the document]
Comment (s-alves10) [reply lib/Url.js]: Moved to ExpensiMark.js

**Comment by s-alves10** [comment]
> Sorry to keep changing my recommendations, I want to make sure this works as well as possible and stays consistent in the future.
> 
> How about we just move [these two functions from App](https://github.com/Expensify/App/blob/0ec61bf1ca783ee0e30fbbc3db7285a33e96744a/src/libs/actions/Report.js#L900-L924) to expensify-common, then we can simply use them in App?

@neil-marcellini 
Please let me know the reason what's better to move those functions here.
Please let me know where you suggest to move those functions to.

**Comment by MariaHCD** [comment]
Looks like @neil-marcellini has got this covered, un-assigning myself as reviewer.

**Comment by s-alves10** [comment]
@neil-marcellini 
Please let me know your thoughts about my comments. I'll make changes according to you.
I hope make this PR merged asap.

**Comment by s-alves10** [comment]
> @s-alves10 I will look at this within the next hour. Please feel free to message me on NewDot if you want to chat quickly.

Please let me know your address on NewDot?

**Review by neil-marcellini — COMMENTED** [review]
We started a group DM in NewDot to discuss.

File: lib/ExpensiMark.js [hunk 4]
```diff
@@ -722,4 +721,31 @@ export default class ExpensiMark {
         // If there are any tags left in the stack, they're unclosed
         return tagStack.length !== 0;
     }
+
+    /**
+     * @param {String} comment
+     * @returns {Array}
+     */
+    extractLinksInMarkdownComment(comment) {
+        const regex = new RegExp(MARKDOWN_LINK_REGEX, 'gi');
```
Comment (neil-marcellini) [inline lib/ExpensiMark.js:730]: Let's declare this at the top of the file.

File: lib/ExpensiMark.js (hunk 1 above)
Comment (neil-marcellini) [inline lib/ExpensiMark.js:6]: Let's do this instead to remove even more code duplication.

```javascript
const MARKDOWN_LINK_REGEXP = new RegExp(`\\[([^\\][]*(?:\\[[^\\][]*][^\\][]*)*)]\\(${MARKDOWN_URL_REGEX}\\)(?![^<]*(<\\/pre>|<\\/code>))`, 'gi');
```

File: lib/ExpensiMark.js [hunk 5]
```diff
@@ -722,4 +721,31 @@ export default class ExpensiMark {
...
+    }
+
+    /**
+     * Compares two markdown comments and returns a list of the links removed in a new comment.
+     *
+     * @param {String} oldComment
+     * @param {String} newComment
+     * @returns {Array}
+     */
+    getRemovedMarkdownLinks(oldComment, newComment) {
```
[the remark this answers is not in the document]
Comment (neil-marcellini) [reply lib/ExpensiMark.js:746]: Expensimark is already exported so we can access it through `Expensimark.getRemovedMarkdownLinks` right?

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Looking pretty good!

File: lib/ExpensiMark.js [hunk 6]
```diff
@@ -73,10 +75,7 @@ export default class ExpensiMark {
             {
                 name: 'link',
                 process: (textToProcess, replacement) => {
-                    const regex = new RegExp(
-                        `\\[([^\\][]*(?:\\[[^\\][]*][^\\][]*)*)]\\(${MARKDOWN_URL_REGEX}\\)(?![^<]*(<\\/pre>|<\\/code>))`,
-                        'gi',
-                    );
+                    const regex = new RegExp(MARKDOWN_LINK_REGEX, 'gi');
                     return this.modifyTextForUrlLinks(regex, textToProcess, replacement);
```
Comment (neil-marcellini) [inline lib/ExpensiMark.js:79]: To clarify, here we can just pass `MARKDOWN_LINK_REGEXP` as the first param

Comment (s-alves10) [reply lib/ExpensiMark.js:79]: like this?
```
    return this.modifyTextForUrlLinks(MARKDOWN_LINK_REGEX, textToProcess, replacement);
```

File: lib/ExpensiMark.js (hunk 4 above)
Comment (s-alves10) [reply lib/ExpensiMark.js]: Done

File: lib/ExpensiMark.js (hunk 6 above)
Comment (s-alves10) [reply lib/ExpensiMark.js]: Done

**Pull request merged**